scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -196,132 +199,159 @@ object Execution {
-     * error to add something to the cache twice clearly).
+     * We send messages from other threads into the submit thread here
      */
-    def getOrElseInsert[T](ex: Execution[T],
-      res: => (EvalCache, Future[(T, ExecutionCounters, EvalCache)]))(implicit ec: ConcurrentExecutionContext): (EvalCache, Future[(T, ExecutionCounters, EvalCache)])
-
-    def ++(that: EvalCache): EvalCache = new EvalCache {
-      def getOrElseInsert[T](ex: Execution[T],
-        res: => (EvalCache, Future[(T, ExecutionCounters, EvalCache)]))(implicit ec: ConcurrentExecutionContext) =
-        that.getOrElseInsert(ex, self.getOrElseInsert(ex, res))
-    }
-  }
-  /**
-   * This is an implementation that remembers history forever.
-   * Since Hadoop jobs are generally long running and not infinite loops,
-   * this is generally safe. If someone wants to make an infinite loop or giant loop,
-   * this may OOM. The solution might be use an immutable LRU cache.
-   */
-  private case class MapEvalCache(cache: Map[Execution[_], Future[(_, ExecutionCounters, EvalCache)]]) extends EvalCache {
-    def getOrElseInsert[T](ex: Execution[T], res: => (EvalCache, Future[(T, ExecutionCounters, EvalCache)]))(implicit ec: ConcurrentExecutionContext) = cache.get(ex) match {
-      case None =>
-        val (next, fut) = res
-        // Make sure ex is added to the cache:
-        val resCache = next ++ MapEvalCache(cache + (ex -> fut))
-        /*
-           * Note in this branch, the future returned includes a
-           * next and the ex -> fut mapping
-           */
-        (resCache, fut.map { case (t, ec, fcache) => (t, ec, resCache ++ fcache) })
-
-      case Some(fut) =>
-        /*
-           * The future recorded here may not itself it it's inner cache
-           * (nothing else is ensuring it). So we make sure the same way we do above
-           */
-        val typedFut = fut.asInstanceOf[Future[(T, ExecutionCounters, EvalCache)]]
-        (this, typedFut.map { case (t, ec, fcache) => (t, ec, this ++ fcache) })
+    sealed trait FlowDefAction
+    case class RunFlowDef(conf: Config,
+      mode: Mode,
+      fd: FlowDef,
+      result: Promise[JobStats]) extends FlowDefAction
+    case object Stop extends FlowDefAction
+    private val messageQueue = new LinkedBlockingQueue[FlowDefAction]()
+    /**
+     * Hadoop and/or cascading has some issues, it seems, with starting jobs
+     * from multiple threads. This thread does all the Flow starting.
+     */
+    private val thread = new Thread(new Runnable {
+      def run() {
+        @annotation.tailrec
+        def go(): Unit = messageQueue.take match {
+          case Stop =>
+            ()
+          case RunFlowDef(conf, mode, fd, promise) =>
+            promise.completeWith(
+              try {
+                ExecutionContext.newContext(conf)(fd, mode).run
+              } catch {
+                // Try our best to complete the future
+                case e: Throwable => Future.failed(e)
```

Comment:
Okay. I got nervous about using NonFatal because I know it catches less than Twitter's version of that does, and if we don't complete that future, we are going to deadlock elsewhere, but in the unlikely event someone does control-c when we are at this stage, we want to exit. I'll change this.